conjure-go-client/httpclient/body_handler.go

```diff
@@ -56,48 +55,47 @@ func (b *bodyMiddleware) RoundTrip(req *http.Request, next http.RoundTripper) (*
 
 // setRequestBody returns a function that should be called once the request has been completed.
 func (b *bodyMiddleware) setRequestBody(req *http.Request) (func(), error) {
-	cleanup := func() {}
-
 	if b.requestInput == nil {
-		return cleanup, nil
+		return func() {}, nil
 	}
 
-	// Special case: if the requestInput is an io.ReadCloser and the requestEncoder is nil,
+	// Special case: if the requestInput is a RequestBody and the requestEncoder is nil,
 	// use the provided input directly as the request body.
-	if bodyReadCloser, ok := b.requestInput.(io.ReadCloser); ok && b.requestEncoder == nil {
-		req.Body = bodyReadCloser
-		// Use the same heuristic as http.NewRequest to generate the "GetBody" function.
-		if newReq, err := http.NewRequest("", "", bodyReadCloser); err == nil {
-			req.GetBody = newReq.GetBody
+	if b.requestEncoder == nil {
+		requestBody, ok := b.requestInput.(RequestBody)
+		if !ok {
+			return nil, werror.ErrorWithContextParams(req.Context(), "requestEncoder is nil but requestInput is not RequestBody",
+				werror.SafeParam("requestInputType", fmt.Sprintf("%T", b.requestInput)))
 		}
-		return cleanup, nil
+		return func() {}, requestBody.setRequestBody(req)
```

Comment:
why is a no-op func returned here?

Reply:
this cleanup function is used to return the byte buffer to the pool after the request is complete, but these binary requests do not use the pool so they get an empty func (but not nil, since `defer cleanup()` would panic)